barretenberg/cpp/src/barretenberg/stdlib/recursion/honk/verifier/protogalaxy_recursive_verifier.test.cpp

```diff
@@ -0,0 +1,319 @@
+    using NativeFoldingVerifier = bb::honk::ProtoGalaxyVerifier_<NativeVerifierInstances>;
+
+    // Helper for getting composer for prover/verifier of recursive (outer) circuit
+    template <typename BuilderT> static auto get_outer_composer()
+    {
+        if constexpr (IsGoblinBuilder<BuilderT>) {
+            return GoblinUltraComposer();
+        } else {
+            return UltraComposer();
+        }
+    }
+
+    /**
+     * @brief Create a non-trivial arbitrary inner circuit, the proof of which will be recursively verified
+     *
+     * @param builder
+     * @param public_inputs
+     * @param log_num_gates
+     */
+    static void create_inner_circuit(InnerBuilder& builder, size_t log_num_gates = 10)
+    {
+        using fr_ct = InnerCurve::ScalarField;
+        using fq_ct = InnerCurve::BaseField;
+        using public_witness_ct = InnerCurve::public_witness_ct;
+        using witness_ct = InnerCurve::witness_ct;
+        using byte_array_ct = InnerCurve::byte_array_ct;
+        using fr = typename InnerCurve::ScalarFieldNative;
+
+        // Create 2^log_n many add gates based on input log num gates
+        const size_t num_gates = 1 << log_num_gates;
+        for (size_t i = 0; i < num_gates; ++i) {
+            fr a = fr::random_element();
+            uint32_t a_idx = builder.add_variable(a);
+
+            fr b = fr::random_element();
+            fr c = fr::random_element();
+            fr d = a + b + c;
+            uint32_t b_idx = builder.add_variable(b);
+            uint32_t c_idx = builder.add_variable(c);
+            uint32_t d_idx = builder.add_variable(d);
+
+            builder.create_big_add_gate({ a_idx, b_idx, c_idx, d_idx, fr(1), fr(1), fr(1), fr(-1), fr(0) });
+        }
+
+        // Define some additional non-trivial but arbitrary circuit logic
+        fr_ct a(public_witness_ct(&builder, fr::random_element()));
+        fr_ct b(public_witness_ct(&builder, fr::random_element()));
+        fr_ct c(public_witness_ct(&builder, fr::random_element()));
+
+        for (size_t i = 0; i < 32; ++i) {
+            a = (a * b) + b + a;
+            a = a.madd(b, c);
+        }
+        pedersen_hash<InnerBuilder>::hash({ a, b });
+        byte_array_ct to_hash(&builder, "nonsense test data");
+        blake3s(to_hash);
+
+        fr bigfield_data = fr::random_element();
+        fr bigfield_data_a{ bigfield_data.data[0], bigfield_data.data[1], 0, 0 };
+        fr bigfield_data_b{ bigfield_data.data[2], bigfield_data.data[3], 0, 0 };
+
+        fq_ct big_a(fr_ct(witness_ct(&builder, bigfield_data_a.to_montgomery_form())), fr_ct(witness_ct(&builder, 0)));
+        fq_ct big_b(fr_ct(witness_ct(&builder, bigfield_data_b.to_montgomery_form())), fr_ct(witness_ct(&builder, 0)));
+
+        big_a* big_b;
+    };
+
+  public:
+    static void SetUpTestSuite() { bb::srs::init_crs_factory("../srs_db/ignition"); }
+
+    static std::shared_ptr<Instance> fold_and_verify(const std::vector<std::shared_ptr<Instance>>& instances,
```

Comment:
(I see now that there's a standalone failure test but I don't see a standalone success test)

Reply:
So the reason why we don't have a standalone folding verifier test (we used to when the decider wasn't implemented) is because we don't have certainty that what we've done is meaningfully correct until we ran the decider. What I mean is that I can give you a garbage accumulator produced from just random scalars and points, I make sure that I set the target sum to be correct, and you can fold a valid circuit on top of it, folding will succeed but that doesn't mean your result is the sandwich of several valid circuits. But only the decider can establish that the sandwich is not valid.

Reply:
I agree that we don't know that the two instances were valid until we "decide" but I don't think this means we shouldn't have a standalone test. The same could be said for an GUH recursive verifier but surely we want independent tests for that. The recursive folding verifier is an independent circuit - to me it makes sense to confirm that we can construct and verify a proof over that circuit. Anyway, you already have a standalone failure test, why not just repurpose that for a "success" test?